Persist search term on lessons page

na2hiro opened the pull request:
On the [lessons page](https://didoesdigital.com/typey-type/lessons), when I search for a term, I see a search result. The problem is, when I click one lesson and go back, the page is shown with empty search box again. It'd be expected that the same search term and result is there.

The solution is to persist the search term state in URL using react-router's useHistory().replace() ([doc](https://v5.reactrouter.com/web/api/history)).

na2hiro commented on src/pages/lessons/components/LessonList.tsx:
```diff
@@ -108,11 +108,30 @@ function filterLessons(searchTerm: string, lessonIndex: LessonIndexEntry[]) {
+  const updateSearchParams = useMemo(() =>
+    debounce((q: string) => {
+      history.replace({ search: `?q=${q}` });
+    }, 100), [history]);
```
Initial commit had an input lag issue: for example, when type with steno like sill -> syllable, 3 deletes and 7 adds were more visible character by character. It's because each keydown additionally required updating URL and parsing it before input can be rendered with updated value.

Now input state is back at the component and URL update is debounced and happens when no input happens after 100ms

didoesdigital commented:
@na2hiro thanks for contributing! Nice work on making this improvement.

That page also has `scrollToHeading` behaviour so that when you visit, for example, <http://localhost:3000/typey-type/lessons/#fundamentals> it scrolls to the heading with that ID. On your branch, when I visited <http://localhost:3000/typey-type/lessons/#fundamentals> it replaced the URL with <http://localhost:3001/typey-type/lessons/?q=> and did not scroll to the Fundamentals heading like it should. Can you please fix this?

I also noticed a lint warning:

```
src/pages/lessons/components/LessonList.tsx
  Line 133:6:  React Hook useEffect has a missing dependency: 'updateSearchParams'. Either include it or remove the dependency array  react-hooks/exhaustive-deps
```

React Hook lint warnings are important to address because bugs often sneak in if left unaddressed. Can you please address this too?

na2hiro commented:
Thanks for taking a look at this!

* I updated so changing of search param preserves hash
  * Tested scroll feature works and URL won't get updated on load
  * Also tested with combined search and hash: http://localhost:3000/typey-type/lessons/?q=number#drills
* I added missing dependency

Could you take a look? Thanks!

didoesdigital commented:
@na2hiro I can't see any changes, perhaps you need to push your latest commits?

na2hiro commented:
Oops, sorry. I've pushed it. Not very used to forked repos...

didoesdigital commented:
@na2hiro no worries about getting used to forked repos!

I've merged the PR! Thanks so much for contributing.

In case it interests you, I made [a few extra changes](https://github.com/didoesdigital/typey-type/compare/57c207fdd1fda1d35bdb2e321f9a49f92215c19b...1afbca5f709e0656a14986e597d40766ab113660) before deploying but the changes are now live! You should be able to see them if you hard refresh the page:

https://didoesdigital.com/typey-type/lessons?q=chao#gerard-nolst-trenit%C3%A9
